File: 283_move_zeroes/main.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
  void moveZeroes(vector<int>& nums) {
    int hollow = 0, current;
    while(nums[hollow]) {
        ++hollow;
    }

    current = hollow + 1;

    while(current < nums.size()) {
      if(!nums[current]) {
        ++current;
        continue;
      }
      nums[hollow++] = nums[current++];
    }

    while(hollow < nums.size()) {
      nums[hollow++] = 0;
    }
  }
};

int main() {
  Solution s;
  vector<int> a{0,1,0,3,12};
  s.moveZeroes(a);

  cout << "--zero-------" << endl;
  for (int i = 0; i < a.size(); ++i) {
    cout << a[i] << endl;
  }
}

File: 283_move_zeroes/main.go
package main

import (
	"fmt"
)

func moveZeroes(nums []int) {
	for i := 0; i < len(nums); i++ {
		if nums[i] != 0 {
			i++
			continue
		}

		for j := i + 1; j < len(nums); j++ {
			if nums[j] != 0 {
				nums[i], nums[j] = nums[j], nums[i]
				break
			}
		}
	}
}

func main() {
	fmt.Println("hello world!")
	nums := []int{0, 1, 0, 3, 12}
	moveZeroes(nums)
	fmt.Println(nums)
}
